Add initd script and method to retrieve the easyfind secret key directly from
the u-boot environment.

Also fixes several issues with upgraded Perl components

Patch by Gordon Bos

--- bubba-easyfind
+++ bubba-easyfind
@@ -0,0 +1,22 @@
+#!/sbin/runscript
+
+NAME=bubba-easyfind
+DAEMON=/usr/bin/twistd
+PIDFILE=/var/run/${NAME}.pid
+TWISTEDSCRIPTFILE=/opt/bubba/bin/bubba-easyfind.tac
+DAEMON_ARGS=" --syslog --pidfile=$PIDFILE --python=$TWISTEDSCRIPTFILE"
+
+# Exit if the package is not installed
+[ -x "" ] || exit 0
+
+start() {
+        ebegin "Starting ${NAME}"
+		start-stop-daemon --start --quiet --make-pidfile --pidfile ${PIDFILE} --background --exec ${DAEMON} $DAEMON_ARGS
+        eend $?
+}
+
+stop() {
+        ebegin "Stopping ${NAME}"
+	        start-stop-daemon --stop --quiet --pidfile ${PIDFILE}
+        eend $?
+}
--- bubba-easyfind.hook
+++ bubba-easyfind.hook
@@ -9,5 +9,5 @@
 else
         # Update easyfind db
         echo "Launching easyfind update script"
-        /usr/lib/web-admin/easyfind.pl
+        /opt/bubba/bin/easyfind.pl
 fi
--- bubba-easyfind.tac
+++ bubba-easyfind.tac
@@ -21,7 +21,7 @@
 from zope.interface import implements
 
 
-LOOP_TIMEOUT = 60  # in seconds the wait time until next query
+LOOP_TIMEOUT = 300  # in seconds the wait time until next query
 
 
 class StringProducer(object):
@@ -88,7 +88,7 @@
 """
 
 application = service.Application("Easyfind Update Service")
-config = ConfigObj('/etc/network/easyfind.conf', configspec=spec.split("\n"))
+config = ConfigObj('/etc/bubba/easyfind.conf', configspec=spec.split("\n"))
 validator = Validator()
 config.validate(validator, copy=True)
 try:
@@ -96,7 +96,6 @@
 except IOError:
     log.err("unable to write easyfind config")
 
-
 def easyfind_ip_changed(data):
     try:
         decoded = json.loads(data)
@@ -121,9 +120,13 @@
 
 
 def easyfind_set_ip(new_ip):
+
     cmdline = parse_cmdline()
     try:
-        key = cmdline['key']
+        if 'key' in cmdline:
+	        key = cmdline['key']
+	else:
+	        key = subprocess.check_output(["/opt/bubba/bin/print-u-boot-env", "-n key"]).strip()
     except KeyError:
         log.err("Unable to retrieve secret key from system")
         return
@@ -138,10 +141,7 @@
         log.err("unable to write easyfind config")
 
     # current WAN interface
-    WAN = subprocess.Popen(
-        ['bubba-networkmanager-cli', 'getwanif'],
-        stdout=subprocess.PIPE
-    ).communicate()[0].strip()
+    WAN = 'eth0'
 
     interface = netifaces.ifaddresses(WAN)
     if netifaces.AF_LINK in interface:
--- print-u-boot-env
+++ print-u-boot-env
@@ -0,0 +1,30 @@
+#!/bin/bash
+#===============================================================
+# This script will set the correct values in /etc/fw_env.config
+# for reading the SPI flash on an Excito B3
+#
+# Required by easyfind if you're using a build-in kernel cmdline
+#===============================================================
+
+
+env_file=/etc/fw_env.config
+my_mtd="/dev/mtd1		0x00000		0x10000		0x10000"
+
+numentries=$(cut -d# -f1 ${env_file} | grep -v "^$" | grep -c .)
+if [ $numentries -gt 1 ]; then
+    env_ok=0
+else
+    env_ok=1
+    env_ok=$(cut -d# -f1 ${env_file} | grep -c "^$my_mtd$")
+fi
+
+if [ $env_ok -lt 1 ]; then
+    >&2 echo "Updating fw_env config file"
+    sed -i 's/^\([^#]..\)/#\1/g' ${env_file}
+    echo "">>${env_file}
+    echo "# Bubba B3 config">>${env_file}
+    echo "$my_mtd">>${env_file}
+fi
+
+fw_printenv $*
+
--- web-admin/bin/easyfind.pl
+++ web-admin/bin/easyfind.pl
@@ -6,10 +6,10 @@
 use CGI;
 use JSON;
 
-use strict;
+#use strict;  # GB: Can't use this with mybubba.org self-signed certificate
 use constant WAN_IF => "eth0";
-use constant EASYFIND_CONF => "/etc/network/easyfind.conf";
-use constant KEY => "/etc/network/bubbakey";
+use constant EASYFIND_CONF => "/etc/bubba/easyfind.conf";
+use constant KEY => "/etc/bubba/bubbakey";
 use constant BOOTARGS => "/proc/cmdline";
 
 sub decode_response {
@@ -31,10 +31,19 @@
  		};
  		
  	} else {
-    	$resp{'error'} = "true";
-    	$resp{'msg'} = "Failed to connect to database server";
-    	print STDERR $resp{'msg'}; 
+	    	$resp{'error'} = "true";
+    		$resp{'msg'} = "Failed to connect to database server";
+	    	print STDERR $resp{'msg'}; 
  	}
+
+	# GB: Fix changed handling of boolean values in decode_json function with new JSON::XS 
+	if ($resp{'error'} !~ /[a-zA-Z]/) {
+		if ($resp{'error'}) {
+			$resp{'error'} = "true";
+		} else {
+			$resp{'error'} = "false";
+		}
+	}
  	return %resp;
 }
 
@@ -91,32 +100,19 @@
 
 }
 
-sub get_extip {
-	use LWP::Simple;
-	my $url = "https://easyfind.excito.org/extip.php"; 
-	my $response = get($url);
-	
-    if ($response) {
-        return $response;
-    } else {
-        print STDERR "Failed to get external ip";
-        return 0;
-    }
-}
-
 sub get_mac {
 
 	my($wtr, $rdr, $err);
 	$err = 1; # we want to dump errors here
 
-	my $pid = open3($wtr,$rdr,$err,"/sbin/ifconfig " . WAN_IF);
+	my $pid = open3($wtr,$rdr,$err,"/bin/ifconfig " . WAN_IF);
 	my @if_a = <$rdr>;
 	my $errmsg = <$err>;
 	waitpid($pid,0);
 
 	chomp(@if_a);
 	my $if = join("\n",@if_a);
-	$if=~ /HWaddr ([\d,a-f,A-F][\d,a-f,A-F]:[\d,a-f,A-F][\d,a-f,A-F]:[\d,a-f,A-F][\d,a-f,A-F]:[\d,a-f,A-F][\d,a-f,A-F]:[\d,a-f,A-F][\d,a-f,A-F]:[\d,a-f,A-F][\d,a-f,A-F])/;
+	$if=~ /ether ([\d,a-f,A-F][\d,a-f,A-F]:[\d,a-f,A-F][\d,a-f,A-F]:[\d,a-f,A-F][\d,a-f,A-F]:[\d,a-f,A-F][\d,a-f,A-F]:[\d,a-f,A-F][\d,a-f,A-F]:[\d,a-f,A-F][\d,a-f,A-F])/;
 	my $mac = $1;
 	if (!$mac) {
 		return 0;
@@ -134,17 +130,27 @@
 	if($$ba{"key"}){	
 		return CGI::escape($$ba{"key"});
 		return ($$ba{"key"});
-	}elsif(-e KEY){
-		open INFILE, "< ".KEY or die "Can't open ".KEY." : $!";
-		@a_key = <INFILE>;
-		close INFILE;
-		$key = @a_key[0];
-		$key=~s/\s//g;
-		return $key;
 	} else {
-		print STDERR ("Error, no keyfile\n");
-		return 0;
-	}	
+		$key = `/opt/bubba/bin/print-u-boot-env -n key 2>/dev/null`;
+		if ( $? == -1 ) {
+			if(-e KEY){
+				open INFILE, "< ".KEY or die "Can't open ".KEY." : $!";
+				@a_key = <INFILE>;
+				close INFILE;
+				$key = @a_key[0];
+			} else {
+				print STDERR ("Error, no keyfile\n");
+				return 0;
+			}
+		}
+		$key=~s/\s//g;
+		if ($key eq "zuwnerrb"){ # GB: default content in keyfile is invalid
+			print STDERR ("accessing u-boot environment failed: $!\n");
+			return 0;
+		} else {
+			return $key;
+		}
+	}
 }
 
 sub print_name {
@@ -175,6 +181,8 @@
 	
 	my $ua = LWP::UserAgent->new;
  	$ua->timeout(2);
+        $ua->ssl_opts( verify_hostname => 0,
+                        SSL_verify_mode => IO::Socket::SSL::SSL_VERIFY_NONE, );
  	
  	# set name on server.
 	my $response = $ua->post('https://easyfind.excito.org/',
@@ -195,6 +203,8 @@
 	
 	my $ua = LWP::UserAgent->new;
  	$ua->timeout(2);
+        $ua->ssl_opts( verify_hostname => 0,
+                        SSL_verify_mode => IO::Socket::SSL::SSL_VERIFY_NONE, );
  	
  	# get record data from server.
 	my $response = $ua->post('https://easyfind.excito.org/',
@@ -274,6 +284,7 @@
 	} else {
 		print STDERR "Easyfind not enabled.\n";
 		$response{'error'} = 'true';
+
 		$response{'msg'} = 'Easyfind not enabled.';
 	}
 }
